### PR TITLE
List consumer groups tool

### LINKED ISSUE #489: `list-consumer-groups` tool
Smallest of the consumer-group analysis tools.
Wraps `admin.listGroups()` to enumerate consumer groups on a cluster.

State and type filters are exposed as optional args from day one — they save a client-side filter pass and let an LLM caller ask "show me only Stable groups" or "only the empty ones" without a second round trip.

## Tool surface

- Name: `list-consumer-groups`.
- Predicate: `kafkaBootstrapOrOAuth`.
- Category: `ToolCategory.Kafka`.
- Annotation: `READ_ONLY`.

## Input schema

```typescript
z.object({
  matchStates: z
    .array(
      z.enum([
        "Unknown",
        "PreparingRebalance",
        "CompletingRebalance",
        "Stable",
        "Dead",
        "Empty",
      ]),
    )
    .nonempty()
    .optional()
    .describe(
      "Restrict to groups currently in these states. Omit to return groups in any state. " +
        'Common filters: ["Stable"] for live consumers, ["Empty"] for orphaned groups.',
    ),
  matchType: z
    .enum(["Classic", "Consumer"])
    .optional()
    .describe(
      "Restrict to groups of this protocol type. Omit to return both types. " +
        '"Consumer" is the new KIP-848 protocol; "Classic" is the legacy generation-based one.',
    ),
  cluster_id: z
    .string()
    .optional()
    .describe(
      "Confluent Cloud logical Kafka cluster ID (lkc-...). Discover via list-clusters.",
    ),
  environment_id: z
    .string()
    .optional()
    .describe(
      "Confluent Cloud environment ID (env-...) that owns the cluster. Discover via list-environments.",
    ),
});
```

The enum literals must stay in lockstep with the `ConsumerGroupStates` and `ConsumerGroupTypes` enums in `@confluentinc/kafka-javascript/types/rdkafka.d.ts`.
Either hand-list them (as above, easier to read in tool descriptions) or derive them via `z.enum(Object.values(ConsumerGroupStates))` — pick at impl time.

`matchType` is a scalar, not an array — there are only two protocol types in the enum, so a multi-value filter has just one meaningful selection (`[type]`) and one degenerate one (`[both]`, equivalent to omitting the filter).
The underlying `admin.listGroups({matchConsumerGroupTypes})` does take an array; the handler wraps `parsed.matchType` in `[parsed.matchType]` before passing it through.

`.nonempty()` on `matchStates` rejects `[]`, which would otherwise be ambiguous (empty filter = "match nothing" or "match everything"?).

## Response shape

```typescript
type ListConsumerGroupsResponse = {
  groups: Array<{
    groupId: string;
    state: string;
    type: string;
    protocolType: string;
    isSimpleConsumerGroup: boolean;
  }>;
  errors: Array<{ message: string; code?: number }>;
};
```

Text summary: `Found N consumer group(s) on cluster <id>.`
When filters are in play, `Found N consumer group(s) (filtered).` keeps it simple — listing the filter values back is noise.

`errors` carries any per-broker `LibrdKafkaError` entries `listGroups` surfaces as non-fatal partial failures.
Surface them — don't swallow.
A successful call with zero entries is `groups: [], errors: []`, not an error.

## Edge cases worth pinning in tests

- Empty cluster.
  `groups: []`, `errors: []`, no error.
- Multiple groups in mixed states.
  All returned without filtering.
- `matchStates: ["Stable"]` filter narrows correctly.
  Only Stable groups in the response.
- `matchStates: ["Empty"]` filter returns groups with zero members (the orphan case).
- `matchStates: []` (empty array).
  Zod-level rejection — `.nonempty()` catches this with a specific message.
- Invalid state name.
  Zod-level rejection citing the valid values.
- `listGroups` returns partial failure (`errors` array non-empty alongside `groups`).
  Surface the errors verbatim; do not synthesize a tool-level error if at least one group came back.
- `listGroups` returns total failure (`groups: []` and `errors` populated).
  Single tool-level error citing the first broker error message.

## Acceptance criteria

- Handler implemented under `src/confluent/tools/handlers/kafka/list-consumer-groups-handler.ts` following the conventions in `.claude/rules/tool-handlers.md`.
- Registered in `tool-name.ts`, `tool-registry.ts`, `index.test.ts`, and `tool-registry.test.ts`'s `EXPECTED_PREDICATES` record.
- Unit tests beside the handler pin every edge case above.
- Integration test under `list-consumer-groups-handler.integration.test.ts` exercises the happy path against a real Confluent Cloud cluster (gated on creds per `.claude/rules/integration-tests.md`).
- CHANGELOG entry.
- README "Kafka tools" section gets a one-line mention.

### PR DESCRIPTION
**Pull request by jlrobins**
## `list-consumer-groups` tool

Wraps `admin.listGroups()` on the native Kafka admin client to enumerate consumer groups on a cluster.

- Predicate `kafkaBootstrapOrOAuth` (same as `consume-messages`, `produce-message`, `list-topics`).
- Category `ToolCategory.Kafka`; annotation `READ_ONLY`.
- Optional `matchStates` (non-empty array of state names) and `matchType` (scalar `Classic` | `Consumer`) save callers a client-side filter pass.
- Standard `cluster_id` / `environment_id` optional args for OAuth connections.

`matchType` is wrapped in a single-element array before being passed through as `matchConsumerGroupTypes`; the underlying `admin.listGroups` accepts an array there but the tool surface deliberately exposes a scalar, since the two-value enum makes a multi-select either redundant or degenerate.

The state and type names surfaced on the response are the TitleCase forms (`Stable`, `Empty`, `Classic`, `Consumer`, …), not the `SCREAMING_SNAKE_CASE` rdkafka enum-key spellings.
librdkafka returns numeric enum values; a forward map translates the Zod-validated input names to numerics on the way in, and a reverse map translates back on the way out.
`Object.values(ConsumerGroupStates)` is intentionally not used — TypeScript's reverse-mapping of numeric enums returns both names and numbers, which Zod would reject.

## `structuredContent` first-mover (alongside `get-partition-offsets`)

This tool surfaces its machine-readable payload via MCP's `structuredContent` channel (per the [2025-11-25 spec](https://modelcontextprotocol.io/specification/2025-11-25/server/tools)).
Adds `BaseToolHandler.createStructuredResponse(text, structured)` next to the existing `createResponse` — additive, no existing handler changes shape.
The text content keeps its human-readable summary role; the broader migration is tracked in #435.

This branch and `feat/480-get-partition-offsets` (still pre-PR) each introduce the same helper independently; whichever lands first wins and the loser drops its copy on rebase.

Wire-format payload:

```typescript
type ListConsumerGroupsResponse = {
  groups: Array<{
    groupId: string;
    state: string;
    type: string;
    protocolType: string;
    isSimpleConsumerGroup: boolean;
  }>;
  errors: Array<{ message: string; code?: number }>;
};
```

Text summary is `Found N consumer group(s).` plain, or `Found N consumer group(s) (filtered).` when any filter is in play.
The original spec carried an `on cluster <id>` clause; dropped because `resolveKafkaClusterArgs` returns `undefined` for direct connections, and the structured payload already carries everything machine-readable.

## Partial- vs total-failure shape

`admin.listGroups` returns `{groups, errors}` where `errors` carries per-broker `LibrdKafkaError` entries surfaced as non-fatal partial failures.
The handler surfaces them verbatim through the `errors` field of the structured payload — partial successes are not collapsed into a tool-level error.

Total failure (`groups: []` and `errors` populated) returns a tool-level `isError: true` response whose text cites the first broker error message.

## Relationships

Closes #489.
Child of #488 (consumer-group analysis tooling).

### Manual testing instructions

1. Prompts like "Show me the consumer groups in cluster X" ("What consumer groups are operating in RealWorldData cluster?") while you have some consumers running:
> One consumer group is active on the realworld-data-cluster (lkc-5vmjd8) in the realworld-data-env environment:
>Group ID | State | Type | Protocol
> -- | -- | -- | --
> 3d8e3c56-3c86-4f3a-9cd0-c82e38037805 | Stable | Classic | consumer

<br class="Apple-interchange-newline">

## Pull request checklist

Please check if your PR fulfills the following (if applicable):

#### Tests

- [x] Added new
- [ ] Updated existing
- [ ] Deleted existing

#### Release notes

- [x] Does anything in this PR need to be mentioned in the user-facing [CHANGELOG](https://github.com/confluentinc/mcp-confluent/blob/main/CHANGELOG.md)?

### CODE REVIEW

File: src/confluent/tools/handlers/kafka/list-consumer-groups-handler.ts [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+  );
+
+const ENUM_TO_TYPE_NAME: ReadonlyMap<KafkaJS.ConsumerGroupTypes, TypeName> =
+  new Map(
+    (
+      Object.entries(TYPE_NAME_TO_ENUM) as Array<
+        [TypeName, KafkaJS.ConsumerGroupTypes]
+      >
+    ).map(([name, value]) => [value, name]),
+  );
```
Comment (bcscc) [inline src/confluent/tools/handlers/kafka/list-consumer-groups-handler.ts:90]: should we add a pin test against the enum in the upstream `KafkaJS.ConsumerGroupStates` in case any changes. Something like:
```
  describe("STATE_NAME_TO_ENUM pin", () => {
    it("should cover every numeric value in KafkaJS.ConsumerGroupStates", () => {
      const upstream = Object.values(KafkaJS.ConsumerGroupStates).filter(
        (v): v is KafkaJS.ConsumerGroupStates => typeof v === "number",
      );
      const mapped = new Set(Object.values(STATE_NAME_TO_ENUM));
      const missing = upstream.filter((v) => !mapped.has(v));
      expect(missing).toEqual([]);
    });
  });
```

File: src/confluent/tools/handlers/kafka/list-consumer-groups-handler.ts [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+
+export const listConsumerGroupsArgs = z.object({
+  matchStates: z
+    .array(z.enum(STATE_NAMES))
+    .nonempty()
+    .optional()
+    .describe(
+      "Restrict to groups currently in these states. Omit to return groups in any state. " +
+        'Common filters: ["Stable"] for live consumers, ["Empty"] for orphaned groups.',
```
Comment (bcscc) [inline src/confluent/tools/handlers/kafka/list-consumer-groups-handler.ts]:
```suggestion
      'Filter to groups in any of these states. Omit to return all. ' +
      'Options: "Stable" (live), "Empty" (orphaned), "Dead", ' +
      '"PreparingRebalance", "CompletingRebalance", "Unknown".',
```

Should we list all options for clarity?

File: src/confluent/tools/handlers/kafka/list-consumer-groups-handler.ts [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    ),
+  matchType: z
+    .enum(TYPE_NAMES)
+    .optional()
+    .describe(
+      "Restrict to groups of this protocol type. Omit to return both types. " +
+        '"Consumer" is the new KIP-848 protocol; "Classic" is the legacy generation-based one.',
```
Comment (bcscc) [inline src/confluent/tools/handlers/kafka/list-consumer-groups-handler.ts]:
```suggestion
      'Filter to groups of this protocol type. Omit to return both. ' +
      'Options: "Consumer", "Classic".',
```

nit/question, is it important to include the additional details for origin of each type?

File: src/confluent/tools/handlers/kafka/list-consumer-groups-handler.ts [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+
+  getToolConfig(): ToolConfig {
+    return {
+      name: ToolName.LIST_CONSUMER_GROUPS,
+      description:
+        "Enumerate consumer groups on a Kafka cluster. Read-only — wraps " +
+        "the broker's listGroups admin call. Optional matchStates / matchType " +
+        "filters narrow the result server-side. Response carries one entry " +
+        "per group ({groupId, state, type, protocolType, isSimpleConsumerGroup}) " +
+        "plus an errors array carrying any per-broker partial failures.",
```
Comment (bcscc) [inline src/confluent/tools/handlers/kafka/list-consumer-groups-handler.ts]:
```suggestion
        "List consumer groups on a Kafka cluster. Optional filters narrow " +
        "the result server-side. Response carries one entry per group with " +
        "groupId, state, type, protocolType, and isSimpleConsumerGroup, plus " +
        "an errors array of any per-broker partial failures.",
```

File: src/confluent/tools/handlers/kafka/list-consumer-groups-handler.ts [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+ * input vocabulary; surface it as the literal `"Unknown"` so callers see a
+ * stable string rather than a number.
+ */
+function typeName(value: KafkaJS.ConsumerGroupTypes): string {
+  if (value === KafkaJS.ConsumerGroupTypes.UNKNOWN) return "Unknown";
+  const name = ENUM_TO_TYPE_NAME.get(value);
+  if (name === undefined) {
+    logger.warn(
+      { value },
+      "Wacky -- unrecognized ConsumerGroupTypes numeric value from librdkafka; surfacing as 'Unknown'",
```
Comment (bcscc) [inline src/confluent/tools/handlers/kafka/list-consumer-groups-handler.ts:245]: I like this error message 😆

File: README.md [hunk 6]
```diff
@@ -321,6 +321,7 @@ get-flink-statement-exceptions: Retrieve the 10 most recent exceptions for a Fli
 get-topic-config: Retrieve configuration details for a specific Kafka topic.
 list-clusters: Get all clusters in the Confluent Cloud environment
 list-connectors: Retrieve a list of "names" of the active connectors. You can then make a read request for a specific connector by name.
+list-consumer-groups: Enumerate consumer groups on a Kafka cluster. Read-only — wraps the broker's listGroups admin call. Optional matchSta...
```
Comment (bcscc) [inline README.md]:
```suggestion
list-consumer-groups: List consumer groups on a Kafka cluster. Read-only — wraps the broker's listGroups admin call. 
```

For consistency

**Comment by bcscc** [comment]
Let's go new tool, thanks @jlrobins, just some minor adjustments & questions

File: src/confluent/tools/handlers/kafka/list-consumer-groups-handler.ts (hunk 1 above)
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/list-consumer-groups-handler.ts:90]: New test written.

File: src/confluent/tools/handlers/kafka/list-consumer-groups-handler.ts (hunk 2 above)
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/list-consumer-groups-handler.ts]: Reworded.

File: README.md (hunk 6 above)
Comment (jlrobins) [reply README.md]: Reworded.

File: src/confluent/tools/handlers/kafka/list-consumer-groups-handler.ts (hunk 3 above)
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/list-consumer-groups-handler.ts]: I think the origin detail earns its space: between "Consumer" and "Classic", a caller without prior Kafka context can't tell which is current. Shortened.

File: src/confluent/tools/handlers/kafka/list-consumer-groups-handler.ts (hunk 4 above)
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/list-consumer-groups-handler.ts]: re-worded.

File: src/confluent/tools/handlers/kafka/list-consumer-groups-handler.ts (hunk 5 above)
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/list-consumer-groups-handler.ts:245]: Is longstanding tradition for 'should-never-be-reached' error handling!

**Review by bcscc — APPROVED** [review]
Thanks!!
